src/haswell/avx2_convert_latin1_to_utf16.cpp

```diff
@@ -0,0 +1,24 @@
+template <endianness big_endian>
+std::pair<const char*, char16_t*> avx2_convert_latin1_to_utf16(const char *latin1_input, size_t len,
+                                    char16_t *utf16_output) {
+  size_t rounded_len = len & ~0xF; // Round down to nearest multiple of 16
+
+  __m256i byteflip = _mm256_setr_epi64x(0x0607040502030001, 0x0e0f0c0d0a0b0809,
+                                        0x0607040502030001, 0x0e0f0c0d0a0b0809);
+  
+  for (size_t i = 0; i < rounded_len; i += 16) {
+    // Load 16 Latin1 characters into a 128-bit register
+    __m128i in = _mm_loadu_si128((__m128i *)&latin1_input[i]);
+    // Zero extend each set of 8 Latin1 characters to 16 16-bit integers
+    __m256i out = _mm256_cvtepu8_epi16(in);
+    if (big_endian) {
+      out = _mm256_shuffle_epi8(out, byteflip);
```

Comment:
Should this be a shift instead of a shuffle ?

Reply:
A shift might be faster but note that the benchmark results above are for the case where big_endian is false. That is, this 'compile time branch' is not taken in the code that we care about (because big endian is uncommon).